### ISSUE TITLE
"Computing the gradient of circuits that return the state is not supported."

### ISSUE THREAD

**Issue by QuantumFall** [PennyLaneAI/pennylane#3948]
### Feature details

To be able to compute gradients w.r.t state output, or at least being able to use qml.math.fidelity(state1,state2).

In the documentation for [qml.math.fidelity](https://docs.pennylane.ai/en/stable/code/api/pennylane.math.fidelity.html), it says that "Numpy, Autograd, Torch, Tensorflow and Jax" are supported. However, when incorporated into my cost function, I get the error stated in the title. 

Currently, the workaround for using any state in my cost function is to output probabilities from my circuit, take their sqrt, and then use those values in the above function. 

I'm guessing the reason for the error is that in general, we cannot ensure that any differentiable cost function using the statevector will return a real value?

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

Example code:

```
import pennylane as qml
from pennylane import numpy as plnp


def testcirc(params, wires = None, nq = None, depth = None):

    # Apply one layer of hadamards
    for nq_ in range(nq):
        qml.Hadamard(wires=nq_)
        
    param_idx = 0
        
    # Apply layers of single qubit rotations and entangling gates
    for l_ in range(depth):
        for i_ in range(nq):
            qml.RY(params[param_idx], wires=i_)
            param_idx += 1
            
        for j in range(0,nq,2):
            if j >= nq-1:
                continue
            else:
                qml.CNOT(wires=[j,j+1])
        for k in range(1,nq,2):
            if k >= nq-1:
                continue
            else:
                qml.CNOT(wires=[k,k+1])
    return qml.state()


def get_plus_state(n):
    statevec = plnp.full(shape = 2**n, fill_value = 1/plnp.sqrt((2**n)))
    return statevec


coeffs = [0.2, -0.543]                                                   # Hamiltonian coefficients 
obs = [qml.PauliZ(0) @ qml.PauliZ(1), qml.PauliX(0) @ qml.PauliX(2)]     # Hamiltonian observables
testH = qml.Hamiltonian(coeffs, obs)                                     # Pennylane Hamiltonian
matrixH = qml.utils.sparse_hamiltonian(testH).real.toarray()             # Hamiltonian in matrix form


nq = len(testH.wires)                                   # Number of qubits
n_measurements = None                                   # Number of shots
ndepth = 4                                              # Depth of circuit
stepsize = 0.1                                          # Optimizer stepsize
nsteps = 10                                             # Optimizer iterations
plus_state = get_plus_state(nq)                         # Generates a |+> state to calculate fidelity with
theta = plnp.array(plnp.random.uniform(0,2*plnp.pi, nq * ndepth), requires_grad = True)

dev = qml.device("lightning.qubit", wires=nq, shots=n_measurements)      # Initialize pennylane device
MyQnode = qml.QNode(testcirc,dev,diff_method='best')
opt = qml.AdagradOptimizer(stepsize=stepsize)                            # Initialize optimizer

def mycf(params, nq, depth, hamiltonian):
    # Cost function to be minimized. 
    # One can think of my fidelity to the |+> state as some constraint
    
    state = MyQnode(params, nq = nq, depth = depth)                      # Gets state from circuit
    expval = plnp.dot(plnp.dot(state, hamiltonian), state)               # Calculates <H>
    F = qml.math.fidelity(state, plus_state)                             # Calculates fidelity abs(<psi|+>)**2
    cf = plnp.abs(expval - F)**2                                         # Cost function is (<H> - F)**2
    return cf


init_cf = mycf(theta, nq, ndepth, matrixH)                               # Calculates initial cf value
print("Begin optimization. Init cf:", init_cf)

for i_ in range(nsteps):
    theta, intermediate_cf  = opt.step_and_cost(mycf, theta, nq = nq, depth = ndepth, hamiltonian = matrixH)
    print(intermediate_cf)
    
print("Opt complete")

```

Workaround: Change the circuit to output probabilities. Take sqrt of the probabilities to obtain statevector with real coefficients. Sample code above can be fixed by replacing the circuit and the cost function with those below:

```
def testcirc(params, wires = None, nq = None, depth = None):

    # Apply one layer of hadamards
    for nq_ in range(nq):
        qml.Hadamard(wires=nq_)
        
    param_idx = 0
        
    # Apply layers of single qubit rotations and entangling gates
    for l_ in range(depth):
        for i_ in range(nq):
            qml.RY(params[param_idx], wires=i_)
            param_idx += 1
            
        for j in range(0,nq,2):
            if j >= nq-1:
                continue
            else:
                qml.CNOT(wires=[j,j+1])
        for k in range(1,nq,2):
            if k >= nq-1:
                continue
            else:
                qml.CNOT(wires=[k,k+1])

    return qml.probs(wires=[i for i in range(nq)])


def mycf(params, nq, depth, hamiltonian):
    # Cost function to be minimized. 
    # One can think of my fidelity to the |+> state as some constraint
    
    state = MyQnode(params, nq = nq, depth = depth)**0.5                 # Gets state from circuit
    expval = plnp.dot(plnp.dot(state, hamiltonian), state)               # Calculates <H>
    F = qml.math.fidelity(state, plus_state)                             # Calculates fidelity abs(<psi|+>)**2
    cf = plnp.abs(expval - F)**2                                         # Cost function is (<H> - F)**2
    return cf
```

**Comment by CatalinaAlbornoz** [comment]
Hi @QuantumFall, thank you for your question! 

What happens here is that `lightning.qubit` doesn't support differentiating qml.state. We have noted it down for improving this in future releases.

In the meantime, there are two ways in which you could fix this issue.

Option 1:
Change the device to use `default.qubit.jax`, and run in analytic mode (i.e., not finite shots).

Option 2:
Use [qml.qinfo.fidelity()](https://docs.pennylane.ai/en/stable/code/api/pennylane.qinfo.transforms.fidelity.html), which is designed to work with QNodes. 

Option 2 is a non-trivial change to what you want because it compares two QNodes and you only have one QNode and one state, but you can hack it into something workable. However option 1 might totally change what you want if the shot information is important.

I hope this can help you fix your problem! And thank you for opening this issue.

**Comment by dwierichs** [comment]
Hi @QuantumFall and @CatalinaAlbornoz,
I just stumbled across this and noticed there is another alternative: The problem stems from the differentiation technique used when `diff_method="best"`. It does not support the differentiation of state vectors because they behave fundamentally different than expectation values, probabilities, or even variances.

However, finite differences are more agnostic to what the function is computing, and setting `diff_method="finite-diff"` also works for the example above :)

Note that finite differences need to be used with the correct shift parameter `h`, which can be passed next to the `diff_method` to the QNode. Which shift parameter is correct will depend _a lot_ on the details, but the code snippet seems to run ok with the default setting.
Hope this helps :)

**Comment by CatalinaAlbornoz** [comment]
Thanks for adding this comment @dwierichs ! It's worth noting that finite differences can end up adding errors since it's an approximate differentiation method instead of an analytic one. However it's worth using it if nothing else works!
